include/seqan3/range/container/small_string.hpp

```diff
@@ -0,0 +1,265 @@
+     */
+    template <size_t N>
+    explicit constexpr small_string(char const (&_lit)[N]) noexcept : small_string{}
+    {
+        static_assert(N <= capacity_ + 1, "Length of string literal exceeds capacity of small_string.");
+        assign(_lit);
+    }
+
+    /*!\brief Construction from char.
+     * \param c The character to construct the small_string for.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    explicit constexpr small_string(char const c) noexcept : small_string{}
+    {
+        assign(1, c);
+    }
+
+    /*!\brief Assign from literal.
+     * \param _lit The literal to assign the string from.
+     *
+     * ### Complexity
+     *
+     * Linear in the size of _lit.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    template <size_t N>
+    constexpr small_string & operator=(char const (&_lit)[N]) noexcept
+    {
+        static_assert(N <= capacity_ + 1, "Length of string literal exceeds capacity of small_string.");
+        assign(_lit);
+        return *this;
+    }
+
+    /*!\brief Assign from literal.
+     * \param _lit The literal to assign the string from.
+     *
+     * ### Complexity
+     *
+     * Linear in the size of _lit.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    template <size_t N>
+    constexpr void assign(char const (&_lit)[N]) noexcept
+    {
+        static_assert(N <= capacity_ + 1, "Length of string literal exceeds capacity of small_string.");
+        assert(_lit[N - 1] == '\0');
+        clear();
+        assign(&_lit[0], &_lit[N-1]);
+    }
+    //!\}
+
+    /*!\name Capacity
+     * \{
+     */
+    //!\brief Returns the maximal size which equals the capacity.
+    constexpr size_type max_size() const noexcept
+    {
+        return capacity_;
+    }
+
+    //!\brief Returns the maximal capacity.
+    constexpr size_type capacity() const noexcept
+    {
+        return capacity_;
+    }
+    //!\}
+
+    /*!\name Element access
+     * \{
+     */
+    /*!\brief Returns the content represented as std::string.
+     *
+     * \returns `std::string` The stored string.
+     *
+     * ### Exceptions
+     *
+     * Strong exception guarantee. No data is modified.
+     *
+     * ### Complexity
+     *
+     * Linear in the size of the string.
+     */
+    std::string string() const
+    {
+        return std::string{this->cbegin(), this->cend()};
+    }
+
+    /*!\brief Returns the content represented as 0-terminated c-style string.
+     *
+     * \returns `const char *` The stored string.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     *
+     * ### Complexity
+     *
+     * Constant.
+     */
+    constexpr const char * c_str() const noexcept
+    {
+        return this->data_.data();
+    }
+    //!\}
+
+    /*!\name Modifiers
+     * \{
+     */
+    //!\copydoc seqan3::small_vector::clear
+    constexpr void clear() noexcept
+    {
+        *this = small_string{};
```

Comment:
Probably not a big thing, but shouldn't it be enough to set size to 0 and [0]=0 ? That's constant the current call is linear...

Reply:
that's how it is done in `small_vector` but here we need to ensure that all positions that are presently not filled are the `\0` character such that when I push back a value the string is automatically correctly null terminated

Reply:
Hm, but this function is linear, while correcting on push_back or insert would be constant...

Even if you do stick with this design, use `assign` instead, because the above costs 2n (the RHS is constructed, 0ed and then copied; there is no move!).

Reply:
I refactored the functions to be constant